barretenberg/cpp/src/barretenberg/circuit_checker/circuit_checker.cpp

```diff
@@ -0,0 +1,190 @@
+    auto values = init_empty_values<Builder>();
+    Params params;
+    params.eta = memory_data.eta; // used in Auxiliary relation for RAM/ROM consistency
+
+    // TODO(https://github.com/AztecProtocol/barretenberg/issues/867): Once we sort gates into their respective blocks
+    // we'll need to either naively run this on all blocks or run only the relevant checks on each block.
+    auto& block = builder.blocks.main;
+
+    // Perform checks on each gate defined in the builder
+    bool result = true;
+    for (size_t idx = 0; idx < block.size(); ++idx) {
+        populate_values(builder, block, values, tag_data, memory_data, idx);
+
+        result = result && check_relation<Arithmetic>(values, params);
+        result = result && check_relation<Elliptic>(values, params);
+        result = result && check_relation<Auxiliary>(values, params);
+        result = result && check_relation<GenPermSort>(values, params);
+        result = result && check_lookup(values, lookup_hash_table);
+        if constexpr (IsGoblinBuilder<Builder>) {
+            result = result && check_relation<PoseidonInternal>(values, params);
+            result = result && check_relation<PoseidonExternal>(values, params);
+        }
+    }
+
+    // Tag check is only expected to pass after all gates have been processed
+    result = result && check_tag_data(tag_data);
+
+    return result;
+};
+
+template <typename Relation> bool CircuitChecker::check_relation(auto& values, auto& params)
+{
+    // Define zero initialized array to store the evaluation of each sub-relation
+    using SubrelationEvaluations = typename Relation::SumcheckArrayOfValuesOverSubrelations;
+    SubrelationEvaluations subrelation_evaluations;
+    for (auto& eval : subrelation_evaluations) {
+        eval = 0;
+    }
+
+    // Evaluate each subrelation in the relation
+    Relation::accumulate(subrelation_evaluations, values, params, /*scaling_factor=*/1);
+
+    // Ensure each subrelation evaluates to zero
+    for (auto& eval : subrelation_evaluations) {
+        if (eval != 0) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool CircuitChecker::check_lookup(auto& values, auto& lookup_hash_table)
+{
+    // If this is a lookup gate, check the inputs are in the hash table containing all table entries
+    if (!values.q_lookup.is_zero()) {
+        return lookup_hash_table.contains({ values.w_l + values.q_r * values.w_l_shift,
+                                            values.w_r + values.q_m * values.w_r_shift,
+                                            values.w_o + values.q_c * values.w_o_shift,
+                                            values.q_o });
+    }
+    return true;
+};
+
+bool CircuitChecker::check_tag_data(const TagCheckData& tag_data)
+{
+    return tag_data.left_product == tag_data.right_product;
+};
+
+template <typename Builder>
+void CircuitChecker::populate_values(
```

Comment:
There is no description what this function does

Reply:
All of the descriptions are in the header. Cody has mentioned that this is better for doxygen for some reason